File: 456-132-pattern/456-132-pattern.cpp
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        return optimised(nums);
        // return solve1(nums);
        // return bruteForce(nums);
    }
private:
    // Daily
    
    // O(n) Time and O(1) space 
    bool optimised(vector<int>& nums) {
        int n = nums.size();
        int top = n, thirdElement = INT_MIN;
        
        for(int i = n - 1; i >= 0; --i) {
            if(thirdElement > nums[i]) return true;
            
            while(top < n && nums[i] > nums[top]) 
                thirdElement = nums[top++];
            nums[--top] = nums[i];
        }
        return false;
    }
    
    
    // O(n^2) Time and O(1) space - TLE
    bool solve1(vector<int>& nums) {
        int n = nums.size();
        int minValue = INT_MAX;
        
        for(int j = 0; j < n; ++j) {
            minValue = min(minValue, nums[j]);
            
            if(minValue == nums[j])
                continue;
            
            for(int k = n - 1; k > j; --k) {
                if(minValue < nums[k] && nums[k] < nums[j])
                        return true;
            }
        }
        return false;
    }    
    
    
    // O(n^3) Time and O(1) space - TLE
    bool bruteForce(vector<int>& nums) {
        int n = nums.size();
        for(int i = 0; i < n - 2; ++i) {
            for(int j = i + 1; j < n - 1; ++j) {
                for(int k = j + 1; k < n; ++k) {
                    if(nums[i] < nums[k] && nums[k] < nums[j])
                        return true;
                }
            }
        }
        return false;
    }
};

File: 456-132-pattern/456-132-pattern.py
class Solution:
    # O(n^3) Time and O(1) Space - TLE
    def __bruteForce(self, nums):
        n = len(nums)
        
        for i in range(n - 2):
            for j in range(i + 1, n - 1):
                for k in range(j + 1, n):
                    if nums[i] < nums[k] < nums[j]:
                        return True
        return False
    
    # O(n^2) Time and O(1) Space - TLE
    def __solve(self, nums):
        n = len(nums)
        
        minElement = float('inf')
        
        for j in range(n):
            minElement = min(minElement, nums[j])
            
            if minElement == nums[j]:
                continue
            
            for k in range(n - 1, j, -1):
                if minElement < nums[k] < nums[j]:
                    return True
        return False
    
    # O(n) Time and O(1) Space
    def __optimised(self, nums):
        n = len(nums)
        top, thirdElement = n, -float('inf')
        
        for i in range(n - 1, -1, -1):
            if thirdElement > nums[i]:
                return True
            
            while top < n and nums[i] > nums[top]:
                thirdElement = nums[top]
                top += 1
            top -= 1
            nums[top] = nums[i]
        return False
        
    def find132pattern(self, nums: List[int]) -> bool:
        return self.__optimised(nums)
        # return self.__solve(nums)
        # return self.__bruteForce(nums)
    
    
    
    
    
    
    